File: 542-01-matrix.js
// see py file for thought process

var updateMatrix = function(mat) {
    let m = mat.length;
    let n = mat[0].length;

    // first pass left to right, top to bottom
    for (let row = 0; row < m; row++) {
        for (let col = 0; col < n; col++) {
            if (mat[row][col] !== 0) {
                let top, left;
                if (row > 0) {
                    top = mat[row-1][col];
                } else {
                    top = Infinity;
                }

                if (col > 0) {
                    left = mat[row][col-1];
                } else {
                    left = Infinity;
                }

                mat[row][col] = Math.min(top, left)+1;
            }
        }
    }

    // second pass right to left and bottom to top
    for (let row = m-1; row >= 0; row--) {
        for (let col = n-1; col >= 0; col--) {
            if (mat[row][col] !== 0) {
                let right,bottom;
                if (col < n-1) {
                    right = mat[row][col+1];
                } else {
                    right = Infinity;
                }

                if (row < m-1) {
                    bottom = mat[row+1][col];
                } else {
                    bottom = Infinity;
                }

                mat[row][col] = Math.min(mat[row][col], Math.min(right, bottom)+1);
            }
        }
    }

    return mat;
};

File: 542-01-matrix.py
# Given an m x n binary matrix mat, return the distance of the nearest 0 for each cell.

# The distance between two adjacent cells is 1.

# Example 1:

# Input: mat = [[0,0,0],[0,1,0],[0,0,0]]
# Output: [[0,0,0],[0,1,0],[0,0,0]]

# Example 2:

# Input: mat = [[0,0,0],[0,1,0],[1,1,1]]
# Output: [[0,0,0],[0,1,0],[1,2,1]]

#############################################################

# dp
# 1.) update values through 2 passes
# 2.) first going left to right and top to bottom
# 3.) second going right to left and bottom to top

class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        m = len(mat)
        n = len(mat[0])

        # first pass go left to right, up to down
        for row in range(m):
            for col in range(n):
                if mat[row][col] != 0:
                    top = mat[row-1][col] if row > 0 else float('inf')
                    left = mat[row][col-1] if col > 0 else float('inf')
                    mat[row][col] = min(top,left)+1
        # second pass go right to left, down to up
        for row in range(m-1, -1, -1):
            for col in range(n-1, -1, -1):
                right = mat[row][col+1] if col < n-1 else float('inf')
                down = mat[row+1][col] if row < m-1 else float('inf')
                mat[row][col] = min(mat[row][col], min(right, down)+1)

        return mat
